lib/android-helpers.js

```diff
@@ -203,49 +204,45 @@ helpers.getDeviceInfoFromCaps = async function getDeviceInfoFromCaps (opts = {})
-      }
-      logger.info(`Looking for a device with Android '${opts.platformVersion}'`);
+      const platformVersion = semver.coerce(opts.platformVersion) || opts.platformVersion;
+      logger.info(`Looking for a device with Android '${platformVersion}'`);
 
       // in case we fail to find something, give the user a useful log that has
       // the device udids and os versions so they know what's available
       const availDevices = [];
       let partialMatchCandidate = null;
-      const extractVersionDigits = (versionStr) => {
-        const match = /(\d+)\.?(\d+)?/.exec(versionStr);
-        return match ? match.slice(1) : [];
-      };
-      const [majorPlatformVersion, minorPlatformVersion] = extractVersionDigits(platformVersion);
       // first try started devices/emulators
       for (const device of devices) {
         // direct adb calls to the specific device
         await adb.setDeviceId(device.udid);
         const rawDeviceOS = await adb.getPlatformVersion();
+        // The device OS could either be a number, like `6.0`
+        // or an abbreviation, like `R`
         availDevices.push(`${device.udid} (${rawDeviceOS})`);
-        const deviceOS = util.coerceVersion(rawDeviceOS, false);
+        const deviceOS = semver.coerce(rawDeviceOS) || rawDeviceOS;
         if (!deviceOS) {
           continue;
         }
 
-        if (util.compareVersions(deviceOS, '==', platformVersion)) {
+        const bothVersionsCanBeCoerced = semver.valid(deviceOS) && semver.valid(platformVersion);
+        const bothVersionsAreStrings = _.isString(deviceOS) && _.isString(platformVersion);
+        if (bothVersionsCanBeCoerced && deviceOS.version === platformVersion.version
+            || bothVersionsAreStrings && _.toLower(deviceOS) === _.toLower(platformVersion)) {
           // Got an exact match - proceed immediately
           udid = device.udid;
           break;
+        } else if (!bothVersionsCanBeCoerced) {
+          // There is no point to check for partial match if either of version numbers is not coercible
+          continue;
         }
 
-        const [majorDeviceVersion, minorDeviceVersion] = extractVersionDigits(deviceOS);
-        if ((!opts.platformVersion.includes('.') && majorPlatformVersion === majorDeviceVersion)
-          || (majorPlatformVersion === majorDeviceVersion && minorPlatformVersion === minorDeviceVersion)) {
-          // Got a partial match - make sure we consider the most recent
-          // device version available on the host system
-          if (partialMatchCandidate
-              && util.compareVersions(deviceOS, '>', _.values(partialMatchCandidate)[0])
-              || !partialMatchCandidate) {
-            partialMatchCandidate = {[device.udid]: deviceOS};
-          }
+        if ((!_.includes(opts.platformVersion, '.') && platformVersion.major === deviceOS.major
+            || platformVersion.major === deviceOS.major && platformVersion.minor === deviceOS.minor)
```

Comment:
Should there be some extra parentheses in here? I'm trying to parse this condition and it is confusing.

Reply:
no parentheses needed, AND has priority over OR. The meaning of this particular expression is:
- if we don't have a dot in platformVersion then only compare major numbers
- otherwise compare both major and minor version numbers
- if any of the conditions above is truthy then bump our candidate if needed ( e.g. the current candidate is not assigned yet or its version is older than the current deviceOS version)